scipy/optimize/_hungarian.py

```diff
@@ -1,92 +1,105 @@
 #
-# Copyright (c) 2008 Brian M. Clapper <bmc@clapper.org>, Gael Varoquaux
-# Author: Brian M. Clapper, Gael Varoquaux
+# Copyright (c) 2019 Jacob Turner <jacob.turner870@gmail.com>
 # License: 3-clause BSD
 
 import numpy as np
-
-
-def linear_sum_assignment(cost_matrix):
-    """Solve the linear sum assignment problem.
-
-    The linear sum assignment problem is also known as minimum weight matching
-    in bipartite graphs. A problem instance is described by a matrix C, where
-    each C[i,j] is the cost of matching vertex i of the first partite set
-    (a "worker") and vertex j of the second set (a "job"). The goal is to find
-    a complete assignment of workers to jobs of minimal cost.
-
-    Formally, let X be a boolean matrix where :math:`X[i,j] = 1` iff row i is
-    assigned to column j. Then the optimal assignment has cost
-
-    .. math::
-        \\min \\sum_i \\sum_j C_{i,j} X_{i,j}
-
-    s.t. each row is assignment to at most one column, and each column to at
-    most one row.
-
-    This function can also solve a generalization of the classic assignment
-    problem where the cost matrix is rectangular. If it has more rows than
-    columns, then not every row needs to be assigned to a column, and vice
-    versa.
-
-    The method used is the Hungarian algorithm, also known as the Munkres or
-    Kuhn-Munkres algorithm.
-
-    Parameters
-    ----------
+from collections import deque, namedtuple
+
+"""
+ Algorithm that solves the minimum / maximum weighted assignment problem
+ (or linear sum assignment problem) for bipartite graphs whose weighted
+ bi-adjacency matrices are not necessarily square.
+
+ Let C be an n x m matrix. Consider a set of pairs M = {(i, j) in [n] x [m]},
+ or equivalently, edges in the bipartite graph. We say that row i is assigned
+ to column j if (i, j) in M. M is a matching if every row is assigned to at
+ most one column and likewise each column is assigned to at most one row.
+
+ The problem is to find a matching M of size min(n, m) such that the sum of
+ C[i,j] for (i, j) in M is minimized / maximized.
+
+ Definitions:
+    Many times we are performing operations on the graph whose edges are
+    designated by zero entries in the assignment matrix and non-zero entries
+    are non-edges. We call this graph the **0-induced graph**.
+
+    If a row has been assigned a column, we say that it is **saturated**. In
+    the algorithm, this is kept track of via the row_saturated and columns
+    saturated vectors.
+
+    Marking is done when we determine a minimum vertex cover of the 0-induced
+    graph. We mark all vertices reachable by augmenting paths from free row
+    vertices. From the resulting set the column vertices are kept and the
+    complement of the row vertices are kept.
+
+ The algorithm consists of six main parts:
+    1. A pre-processing of the input matrix for easy access in the main
+       algorithms
+    2. Adjust weights so that each row has a zero entry. (Step 1 Wikipedia)
```

Comment:
is there an article with steps 1-4 you can reference? A general reference to Wikipedia is fine, but specific steps there can change over time.